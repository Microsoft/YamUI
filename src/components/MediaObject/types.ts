/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
export enum MediaObjectSize {
  XLARGE = 'xLarge',
  LARGE = 'large',
  MEDIUM = 'medium',
  SMALL = 'small',
  XSMALL = 'xSmall',
}
