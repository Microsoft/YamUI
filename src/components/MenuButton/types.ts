/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { ContextualMenuItemType } from 'office-ui-fabric-react/lib/ContextualMenu';

export enum MenuItemType {
  Normal = ContextualMenuItemType.Normal,
  Header = ContextualMenuItemType.Header,
  Divider = ContextualMenuItemType.Divider,
}
