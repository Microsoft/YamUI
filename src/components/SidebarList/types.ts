/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { BaseIllustration } from '../Illustration';

export type SidebarListSize = 'small' | 'medium' | 'large';

export interface BaseSidebarListItem {
  /**
   * A unique key for the item.
   */
  key: string;

  /**
   * The visible title text.
   */
  title: string;

  /**
   * Metadata text displayed under the title, not displayed at small size.
   */
  description?: string;
}
export interface SidebarListIllustrationItem extends BaseSidebarListItem {
  /**
   * A YamUI Illustration component class. You are responsible for importing and providing the correct size.
   */
  illustration: typeof BaseIllustration;
}
export interface SidebarListImageItem extends BaseSidebarListItem {
  /**
   * Custom image JSX. It should fit the list item's pixel width at the given size, or fill 100% of its container width.
   */
  imageContent: React.ReactNode;
}
export type SidebarListItemWithGraphic = SidebarListIllustrationItem | SidebarListImageItem;

export type SidebarListClickableItem = SidebarListItemWithGraphic & {
  /**
   * Callback to be triggered when the item is clicked/invoked.
   */
  onClick(key: BaseSidebarListItem['key']): void;
};
export type SidebarListLinkItem = SidebarListItemWithGraphic & {
  /**
   * The URL this item links to.
   */
  href: string;

  /**
   * If true, opens the link securely in a new window.
   */
  newWindow?: boolean;
};
export type SidebarListItem = SidebarListClickableItem | SidebarListLinkItem;
