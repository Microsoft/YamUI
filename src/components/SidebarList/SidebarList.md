### Examples

Small size with links to documents:

```js { "props": { "data-description": "small", "data-action-states": "[{\"action\":\"none\"},{\"action\":\"hover\",\"selector\":\".y-sidebarList > li:nth-child(2) .y-sidebarList--item\"}]" } }
const Docx32 = require('../Illustration/illustrations/Docx32').default;
const One32 = require('../Illustration/illustrations/One32').default;
const Xlsx32 = require('../Illustration/illustrations/Xlsx32').default;
const Pptx32 = require('../Illustration/illustrations/Pptx32').default;

const items = [
  {
    key: '1',
    title: 'Financial outlook',
    illustration: Docx32,
    href: "https://www.yammer.com",
    newWindow: true,
  },
  {
    key: '2',
    title: 'Goals for Q3',
    illustration: One32,
    href: "https://www.yammer.com",
    newWindow: true,
  },
  {
    key: '3',
    title: 'Previous year financial records',
    illustration: Xlsx32,
    href: "https://www.yammer.com",
    newWindow: true,
  },
  {
    key: '4',
    title: 'AllHands financial slides',
    illustration: Pptx32,
    href: "https://www.yammer.com",
    newWindow: true,
  },
];

<SidebarList items={items} size="small" />
```

Medium size with links to related groups:

```js { "props": { "data-description": "medium", "data-action-states": "[{\"action\":\"none\"},{\"action\":\"hover\",\"selector\":\".y-sidebarList > li:nth-child(2) .y-sidebarList--item\"}]" } }
const { BorderType, ImageFit } = require('../Image');

const items = [
  {
    key: '1',
    title: 'Model Rocket Enthusiasts',
    imageContent: <Image source="group.png" description="Model Rocket Enthusiasts group logo" width={24} height={24} borderType={BorderType.SOFT} imageFit={ImageFit.cover} block={true} />,
    href: "https://www.yammer.com",
    newWindow: true,
  },
  {
    key: '2',
    title: 'Super Fun Happy Times Group',
    imageContent: <Image source="user.png" description="Super Fun Happy Times Group group logo" width={24} height={24} borderType={BorderType.SOFT} imageFit={ImageFit.cover} block={true} />,
    href: "https://www.yammer.com",
    newWindow: true,
  },
  {
    key: '3',
    title: 'Product Management',
    imageContent: <Image source="logo.png" description="Product Management group logo" width={24} height={24} borderType={BorderType.SOFT} imageFit={ImageFit.cover} block={true} />,
    href: "https://www.yammer.com",
    newWindow: true,
  },
  {
    key: '4',
    title: 'Engineering',
    imageContent: <Image source="logoFallback.png" description="Engineering group logo" width={24} height={24} borderType={BorderType.SOFT} imageFit={ImageFit.cover} block={true} />,
    href: "https://www.yammer.com",
    newWindow: true,
  },
];

<SidebarList items={items} size="medium" />
```

Large size with links to related groups:

```js { "props": { "data-description": "large", "data-action-states": "[{\"action\":\"none\"},{\"action\":\"hover\",\"selector\":\".y-sidebarList > li:nth-child(2) .y-sidebarList--item\"}]" } }
const { AvatarSize } = require('../Avatar');

const items = [
  {
    key: '1',
    title: user.name,
    description: user.jobTitle,
    imageContent: <Avatar size={AvatarSize.SMALL} name={user.name} imageUrl={user.imageUrl} />,
    href: "https://www.yammer.com",
    newWindow: true,
  },
  {
    key: '2',
    title: user.name,
    description: user.jobTitle,
    imageContent: <Avatar size={AvatarSize.SMALL} name={user.name} imageUrl={user.imageUrl} />,
    href: "https://www.yammer.com",
    newWindow: true,
  },
  {
    key: '3',
    title: user.name,
    description: user.jobTitle,
    imageContent: <Avatar size={AvatarSize.SMALL} name={user.name} imageUrl={user.imageUrl} />,
    href: "https://www.yammer.com",
    newWindow: true,
  },
  {
    key: '4',
    title: user.name,
    description: user.jobTitle,
    imageContent: <Avatar size={AvatarSize.SMALL} name={user.name} imageUrl={user.imageUrl} />,
    href: "https://www.yammer.com",
    newWindow: true,
  },
];

<SidebarList items={items} size="large" />
```

Items can provide <strong>href</strong> to render as a link, or <strong>onClick</strong> to render as a button:

```js { "props": { "data-description": "interaction options" } }
const Docx32 = require('../Illustration/illustrations/Docx32').default;
const One32 = require('../Illustration/illustrations/One32').default;
const Xlsx32 = require('../Illustration/illustrations/Xlsx32').default;

const items = [
  {
    key: '1',
    title: 'This is a link',
    illustration: Docx32,
    href: "https://www.yammer.com",
  },
  {
    key: '2',
    title: 'This is a link which opens in a new window',
    illustration: One32,
    href: "https://www.yammer.com",
    newWindow: true,
  },
  {
    key: '3',
    title: 'This is a button',
    illustration: Xlsx32,
    onClick: action('clicked!'),
  },
];

<SidebarList items={items} size="small" />
```