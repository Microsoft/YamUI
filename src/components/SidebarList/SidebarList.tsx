/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { join } from '../../util/classNames';
import { BaseComponentProps } from '../../util/BaseComponent/props';
import { LayoutList } from '../LayoutList';
import SidebarListItemView from './SidebarListItem';
import { SidebarListItem, SidebarListSize } from './types';

export interface SidebarListProps extends BaseComponentProps {
  /**
   * A list of item objects which will be rendered as list items.
   */
  items: SidebarListItem[];

  /**
   * "small" will display the title next to a 24px image and compact vertical padding.
   * "medium" will display the title next to a 24px image and slightly more vertical padding.
   * "large" will display the title and description next to a 32px image.
   */
  size: SidebarListSize;

  /**
   * Called when an item is clicked/invoked.
   */
  onClick?(key: SidebarListItem['key']): void;
}

/**
 * SidebarList renders a list of MediaObjects wrapped in either a link or button. This component should be used to list
 * relevant metadata like related groups, pinned documents, group memberships, etc.
 */
export default class SidebarList extends React.Component<SidebarListProps> {
  public render() {
    return (
      <LayoutList direction="vertical" className={join(['y-sidebarList', this.props.className])}>
        {this.props.items.map(this.getItem)}
      </LayoutList>
    );
  }

  private getItem = (item: SidebarListItem) => {
    return <SidebarListItemView item={item} size={this.props.size} key={item.key} />;
  };
}
