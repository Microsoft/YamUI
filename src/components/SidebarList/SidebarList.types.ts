/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { BaseIllustration } from '../Illustration';
import { BaseComponentProps } from '../../util/BaseComponent/props';

export type SidebarListSize = 'small' | 'medium' | 'large';

export interface BaseSidebarListItem {
  /**
   * A unique key for the item.
   */
  key: string;

  /**
   * The visible title text.
   */
  title: string;

  /**
   * Metadata text displayed under the title, not displayed at small size.
   */
  description?: string;
}
export interface SidebarListIllustrationItem extends BaseSidebarListItem {
  /**
   * A YamUI Illustration component class. You are responsible for importing and providing the correct size.
   */
  illustration: typeof BaseIllustration;
}
export interface SidebarListImageItem extends BaseSidebarListItem {
  /**
   * Custom image JSX. It should fit the list item's pixel width at the given size, or fill 100% of its container width.
   */
  imageContent: React.ReactNode;
}
export type SidebarListItemWithGraphic = SidebarListIllustrationItem | SidebarListImageItem;

export type SidebarListClickableItem = SidebarListItemWithGraphic & {
  /**
   * Callback to be triggered when the item is clicked/invoked.
   */
  onClick(key: BaseSidebarListItem['key']): void;
};
export type SidebarListLinkItem = SidebarListItemWithGraphic & {
  /**
   * The URL this item links to.
   */
  href: string;

  /**
   * If true, opens the link securely in a new window.
   */
  newWindow?: boolean;
};
export type SidebarListItem = SidebarListClickableItem | SidebarListLinkItem;

export interface SidebarListProps extends BaseComponentProps {
  /**
   * A list of item objects which will be rendered as list items.
   */
  items: SidebarListItem[];

  /**
   * "small" will display the title next to a 24px image and compact vertical padding.
   * "medium" will display the title next to a 24px image and slightly more vertical padding.
   * "large" will display the title and description next to a 32px image.
   */
  size: SidebarListSize;

  /**
   * Called when an item is clicked/invoked.
   */
  onClick?(key: SidebarListItem['key']): void;
}

export interface SidebarListItemProps {
  item: SidebarListItem;
  size: SidebarListProps['size'];
}
