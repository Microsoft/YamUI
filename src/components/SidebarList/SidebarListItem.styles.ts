/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { mergeStyleSets, ITheme } from 'office-ui-fabric-react/lib/Styling';
import { memoizeFunction } from 'office-ui-fabric-react/lib/Utilities';
import { getGutterValue } from '../../util/styles/gutters';
import { SidebarListSize } from './SidebarList.types';

const getPadding = (size: SidebarListSize) => {
  if (size === 'small') {
    return `${getGutterValue(0.5)} 0`;
  }
  if (size === 'medium') {
    return `${getGutterValue(1)} 0`;
  }
  return `0 0 ${getGutterValue(1)}`;
};

export interface SidebarClassNameProps {
  size: SidebarListSize;
  theme: ITheme;
}

export const getClassNames = memoizeFunction((classNameProps: SidebarClassNameProps) => {
  const { theme, size } = classNameProps;
  return mergeStyleSets({
    root: {
      position: 'relative',
      selectors: {
        ':hover': {
          backgroundColor: theme.semanticColors.menuItemBackgroundHovered,
          margin: `0 ${getGutterValue(-2)}`,
          padding: `0 ${getGutterValue(2)}`,
        },
      },
    },
    content: {
      padding: getPadding(size),
    },
  });
});
