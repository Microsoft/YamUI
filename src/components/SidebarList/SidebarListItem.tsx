/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { join } from '../../util/classNames';
import Clickable from '../Clickable';
import { LayoutListItem } from '../LayoutList';
import MediaObject, { MediaObjectSize } from '../MediaObject';
import NavigationLink from '../NavigationLink';
import { SidebarListProps } from './SidebarList';
import {
  SidebarListItem,
  SidebarListLinkItem,
  SidebarListIllustrationItem,
  SidebarListImageItem,
  SidebarListClickableItem,
} from './types';
import { getClassNames, SidebarListItemClassNames } from './SidebarListItem.styles';

export { SidebarListItem };

const getIllustrationSize = (size: SidebarListItemProps['size']) => (size === 'large' ? 32 : 24);
const getMediaObjectSize = (size: SidebarListItemProps['size']) =>
  size === 'large' ? MediaObjectSize.SMALL : MediaObjectSize.XSMALL;

const isLinkItem = (item: SidebarListItem): item is SidebarListLinkItem => {
  return (item as SidebarListLinkItem).href !== undefined;
};

const getImageContent = (item: SidebarListItem, size: SidebarListItemProps['size']) => {
  if ((item as SidebarListIllustrationItem).illustration) {
    const Illustration = (item as SidebarListIllustrationItem).illustration;
    return <Illustration size={getIllustrationSize(size)} block={true} />;
  }

  return (item as SidebarListImageItem).imageContent;
};

export interface SidebarListItemProps {
  item: SidebarListItem;
  size: SidebarListProps['size'];
}

export default class SidebarListItemView extends React.Component<SidebarListItemProps> {
  public render() {
    const { root, content } = getClassNames(this.props.size);
    return (
      <LayoutListItem>
        <div className={join([root, 'y-sidebarList--item'])}>{this.getItemContent(content)}</div>
      </LayoutListItem>
    );
  }

  private getItemContent = (contentClassName: SidebarListItemClassNames['content']) => {
    const { item, size } = this.props;

    const content = (
      <div className={contentClassName}>
        <MediaObject
          size={getMediaObjectSize(size)}
          imageContent={getImageContent(item, size)}
          titleContent={item.title}
          metadataContent={item.description}
        />
      </div>
    );

    if (isLinkItem(item)) {
      return (
        <NavigationLink href={item.href} newWindow={item.newWindow} block={true} unstyled={true}>
          {content}
        </NavigationLink>
      );
    }

    return (
      <Clickable onClick={this.onClick} block={true} unstyled={true}>
        {content}
      </Clickable>
    );
  };

  private onClick = () => {
    (this.props.item as SidebarListClickableItem).onClick(this.props.item.key);
  };
}
