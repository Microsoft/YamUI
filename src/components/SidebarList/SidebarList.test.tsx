/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import AccDB16 from '../Illustration/illustrations/AccDB16';
import SidebarList, { SidebarListProps } from './index';
import { SidebarListItem } from './SidebarList.types';

describe('<SidebarList />', () => {
  let component: ShallowWrapper<SidebarListProps>;

  describe('with default options and empty items', () => {
    beforeEach(() => {
      component = shallow(<SidebarList items={[]} size="small" />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with additional className', () => {
    beforeEach(() => {
      component = shallow(<SidebarList items={[]} size="large" className="TEST_CLASSNAME" />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with items', () => {
    beforeEach(() => {
      const items: SidebarListItem[] = [
        {
          key: '1',
          title: 'Item 1 title',
          href: 'https://www.yammer.dev',
          illustration: AccDB16,
        },
      ];

      component = shallow(<SidebarList items={items} size="large" />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
