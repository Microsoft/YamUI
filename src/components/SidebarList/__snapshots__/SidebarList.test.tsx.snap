// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SidebarList /> with additional className matches its snapshot 1`] = `
<LayoutList
  className="y-sidebarList TEST_CLASSNAME"
  direction="vertical"
/>
`;

exports[`<SidebarList /> with default options and empty items matches its snapshot 1`] = `
<LayoutList
  className="y-sidebarList"
  direction="vertical"
/>
`;

exports[`<SidebarList /> with items matches its snapshot 1`] = `
<LayoutList
  className="y-sidebarList"
  direction="vertical"
>
  <CustomizedSidebarListItemView
    item={
      Object {
        "href": "https://www.yammer.dev",
        "illustration": [Function],
        "key": "1",
        "title": "Item 1 title",
      }
    }
    key="1"
    size="large"
  />
</LayoutList>
`;
